content/blog/rails-restapi-outside-in-tdd.md

```diff
@@ -0,0 +1,740 @@
+            create_new_todo.create
+        end
+
+$rspec
+
+RED:
+ 1) Todos Controller invokes the instance of Create_Runner successfully and renders the json
+     Failure/Error: allow(create_new_todo).to receive(:json_result).and_return(true)
+       #<TodosController:0x000000000069f0> does not implement: json_result
+GREEN:
+
+    def create
+        result, status = @runner.create_todo(todo_params)
+        json_result(result, status)
+    end
+    def json_result(object, status)
+        render json: object, status:status
+    end
+
+$ rspec
+
+All Greens
+
+```
+
+I went on to write the request spec(todos_spec.rb) that expects 422 when the title, which is a mandatory field, is not passed.
+
+```ruby
+Failing Test Case:
+  context 'with invalid request attributes ' do
+    it 'returns status code 422' do
+      post '/qwerts', params: {title: "", description:'First todo' }, as: :json
+      expect(response).to have_http_status(422)
+    end
+  end
+
+$ rspec
+
+RED:
+  1) Todos with invalid request attributes returns status code 422
+     Failure/Error: expect(response).to have_http_status(422)
+       expected the response to have status code 422 but it was 201
+```
+
+```ruby
+The fix has to be done at Runner layer as this layer contains the application logic.
+todos_runner_spec.rb
+Failing Test Case:
+        it " returns unprocessable entity when title is not passed" do
+            save_params = {title: "", description:'First todo' }
+            todo = Todo.new(save_params)
+            todosRunner = TodosRunner.new
+            todo, status = todosRunner.create_todo(save_params)
+            expect(status).to eq(:unprocessable_entity)
+            expect(todo["message"]).to eq("Invalid input values")
+        end
+Fix:
+    def create_todo(add_params)
+        todo =@repo.new_todo(add_params)
+        if(@repo.save_todo(todo))
+            return todo, :created
+        else
+            error = Hash.new
+            error["message"] = " Invalid input values"
+            return error ,:unprocessable_entity
+        end
+    end
+ $ rspec
+
+All Green
+
+```
+
+### It is always quite a sight to see all green
+
+I was so immersed into conversing with rspec to develop this code, I felt like I was a part of a play, playing a character and that the play is being observed, and I am responding only when there is a cue and refrained to when there is no cue as I may mislead them observers if I take any decision without the cue from the rspec.
+
+In this manner I finished the assignment and gave it to the client.
+I had two days to go for a visit around the place.
+As I was watching a tomb my instinctive thoughts about the plausible errors in my work started to take over me but with the fact that there never existed any piece of code before going through a test did not let those thoughts take over me, and they seem to have evaporated.
+The sightseeing was pleasant with nothing in the back of my mind.
```

Comment:
Now, with my mind at ease and feeling a sense of achievement, the sights all around seemed a bit more pleasant!